src/rust/engine/src/rule_graph.rs

```diff
@@ -454,6 +463,99 @@ pub struct RuleGraph {
   unfulfillable_rules: UnfulfillableRuleMap,
 }
 
+fn type_constraint_str(externs: &Externs, type_constraint: TypeConstraint) -> String {
+  let val = to_val(externs, type_constraint);
+  repr_of_val(externs, &val)
+}
+
+fn to_val(externs: &Externs, type_constraint: TypeConstraint) -> Value {
+  to_val_from_id(externs, type_constraint.0)
+}
+
+fn to_val_from_func(externs: &Externs, func: &Function) -> Value {
+  to_val_from_id(externs, func.0)
+}
+
+fn to_val_from_id(externs: &Externs, id: Id) -> Value {
+  externs.val_for_id(id)
+}
+
+fn repr_of_val(externs: &Externs, value: &Value) -> String {
+  let rpr_val = externs.project_ignoring_type(&value, "__repr__");
+
+  let invoke_result  = externs.invoke_runnable(&Runnable{func: rpr_val, args:vec![], cacheable:false})
+                              .expect("string from calling repr");
+  externs.val_to_str(&invoke_result)
+}
+
+
+fn function_str(externs: &Externs, func: &Function) -> String {
+  let as_val = to_val_from_func(externs, func);
+  val_name(externs, &as_val)
+}
+
+
+fn type_str(externs: &Externs, type_id: TypeId) -> String {
+  if type_id == ANY_TYPE {
```

Comment:
This makes me fidgety, as I'm certain that we will forget to do it in a bunch of places. Would be great to remove this singleton by removing the typeid from `Value` (prolly after both this and #4261 land).